packages/firestore/test/unit/local/document_overlay_cache.test.ts

```diff
@@ -0,0 +1,257 @@
+ * limitations under the License.
+ */
+import { expect } from 'chai';
+
+import { User } from '../../../src/auth/user';
+import { IndexedDbPersistence } from '../../../src/local/indexeddb_persistence';
+import { Persistence } from '../../../src/local/persistence';
+import { documentKeySet, DocumentKeySet } from '../../../src/model/collections';
+import { DocumentKey } from '../../../src/model/document_key';
+import { Mutation, mutationEquals } from '../../../src/model/mutation';
+import { Overlay } from '../../../src/model/overlay';
+import { addEqualityMatcher } from '../../util/equality_matcher';
+import {
+  deleteMutation,
+  key,
+  patchMutation,
+  path,
+  setMutation
+} from '../../util/helpers';
+
+import * as persistenceHelpers from './persistence_test_helpers';
+import { TestDocumentOverlayCache } from './test_document_overlay_cache';
+
+let persistence: Persistence;
+let overlayCache: TestDocumentOverlayCache;
+
+describe('MemoryDocumentOverlayCache', () => {
+  beforeEach(() => {
+    return persistenceHelpers.testMemoryEagerPersistence().then(p => {
+      persistence = p;
+    });
+  });
+
+  genericDocumentOverlayCacheTests();
+});
+
+describe('IndexedDbDocumentOverlayCache', () => {
+  if (!IndexedDbPersistence.isAvailable()) {
+    console.warn('No IndexedDB. Skipping IndexedDbMutationQueue tests.');
+    return;
+  }
+
+  beforeEach(() => {
+    return persistenceHelpers.testIndexedDbPersistence().then(p => {
+      persistence = p;
+    });
+  });
+
+  genericDocumentOverlayCacheTests();
+});
+
+/**
+ * Defines the set of tests to run against both document overlay cache
+ * implementations.
+ */
+function genericDocumentOverlayCacheTests(): void {
+  addEqualityMatcher();
+
+  beforeEach(() => {
+    overlayCache = new TestDocumentOverlayCache(
+      persistence,
+      persistence.getDocumentOverlay(new User('user'))
+    );
+  });
+
+  afterEach(async () => {
+    await persistence.shutdown();
+    await persistenceHelpers.clearTestPersistence();
+  });
+
+  function saveOverlaysForMutations(
+    largestBatch: number,
+    ...mutations: Mutation[]
+  ): Promise<void> {
+    const data: Map<DocumentKey, Mutation> = new Map<DocumentKey, Mutation>();
+    for (const mutation of mutations) {
+      data.set(mutation.key, mutation);
+    }
+    return overlayCache.saveOverlays(largestBatch, data);
+  }
+
+  function saveOverlaysForKeys(
+    largestBatch: number,
+    ...overlayKeys: string[]
+  ): Promise<void> {
+    const data: Map<DocumentKey, Mutation> = new Map<DocumentKey, Mutation>();
+    for (const overlayKey of overlayKeys) {
+      data.set(key(overlayKey), setMutation(overlayKey, {}));
+    }
+    return overlayCache.saveOverlays(largestBatch, data);
+  }
+
+  function verifyEqualMutations(
+    m1: Mutation | null,
+    m2: Mutation | null
+  ): void {
+    expect(m1 === null).to.equal(false);
+    expect(m2 === null).to.equal(false);
+    expect(mutationEquals(m1!, m2!)).to.equal(true);
```

Comment:
Surprised that `m1 === null && m2 === null` returns false;

Reply:
It doesn't expect any null values for neither m1/m2 (line 111/112). But better to update it anyway. Done.